services/oauth/src/main/java/org/mule/services/oauth/internal/AbstractOAuthDancer.java

```diff
@@ -0,0 +1,273 @@
+import static org.mule.service.http.api.HttpConstants.Method.POST;
+import static org.mule.service.http.api.HttpHeaders.Names.AUTHORIZATION;
+import static org.mule.service.http.api.HttpHeaders.Names.CONTENT_TYPE;
+import static org.mule.service.http.api.HttpHeaders.Values.APPLICATION_X_WWW_FORM_URLENCODED;
+import static org.mule.service.http.api.utils.HttpEncoderDecoderUtils.decodeUrlEncodedBody;
+import static org.mule.service.http.api.utils.HttpEncoderDecoderUtils.encodeString;
+
+import org.mule.runtime.api.el.BindingContext;
+import org.mule.runtime.api.el.ExpressionEvaluator;
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.api.lifecycle.Startable;
+import org.mule.runtime.api.lifecycle.Stoppable;
+import org.mule.runtime.api.lock.LockFactory;
+import org.mule.runtime.api.metadata.DataType;
+import org.mule.runtime.api.metadata.MediaType;
+import org.mule.runtime.api.metadata.TypedValue;
+import org.mule.runtime.core.util.IOUtils;
+import org.mule.runtime.core.util.MapUtils;
+import org.mule.runtime.oauth.api.exception.RequestAuthenticationException;
+import org.mule.runtime.oauth.api.exception.TokenNotFoundException;
+import org.mule.runtime.oauth.api.exception.TokenUrlResponseException;
+import org.mule.runtime.oauth.api.state.ResourceOwnerOAuthContext;
+import org.mule.service.http.api.client.HttpClient;
+import org.mule.service.http.api.domain.ParameterMap;
+import org.mule.service.http.api.domain.entity.ByteArrayHttpEntity;
+import org.mule.service.http.api.domain.entity.InputStreamHttpEntity;
+import org.mule.service.http.api.domain.message.request.HttpRequest;
+import org.mule.service.http.api.domain.message.request.HttpRequestBuilder;
+import org.mule.service.http.api.domain.message.response.HttpResponse;
+import org.mule.services.oauth.internal.state.TokenResponse;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.locks.Lock;
+
+public abstract class AbstractOAuthDancer implements Startable, Stoppable {
+
+  private static final int TOKEN_REQUEST_TIMEOUT_MILLIS = 60000;
+
+  protected final String clientId;
+  protected final String clientSecret;
+  protected final String tokenUrl;
+  protected final Charset encoding;
+  protected final String scopes;
+
+  protected final String responseAccessTokenExpr;
+  protected final String responseRefreshTokenExpr;
+  protected final String responseExpiresInExpr;
+  protected final Map<String, String> customParametersExtractorsExprs;
+
+  private final LockFactory lockProvider;
+  private final Map<String, ResourceOwnerOAuthContext> tokensStore;
+  private final HttpClient httpClient;
+  private final ExpressionEvaluator expressionEvaluator;
+
+  protected AbstractOAuthDancer(String clientId, String clientSecret, String tokenUrl, Charset encoding, String scopes,
```

Comment:
Only the subclasses use this constuctor, and that instantiation is made from a builder.